File: day5/part1/main.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_RANGES 1000
#define MAX_LINE 100

typedef struct {
  long long start;
  long long end;
} Range;

int main(void) {
  FILE *file = fopen("list.txt", "r");
  if (!file) {
    perror("Failed to open list.txt");
    return 1;
  }

  Range ranges[MAX_RANGES];
  int range_count = 0;
  char line[MAX_LINE];
  bool parsing_ranges = true;
  int fresh_count = 0;

  while (fgets(line, sizeof(line), file)) {

    line[strcspn(line, "\n")] = '\0';

    if (strlen(line) == 0) {
      parsing_ranges = false;
      continue;
    }

    if (parsing_ranges) {

      long long start, end;
      if (sscanf(line, "%lld-%lld", &start, &end) == 2) {
        if (range_count < MAX_RANGES) {
          ranges[range_count].start = start;
          ranges[range_count].end = end;
          range_count++;
        }
      }
    } else {

      long long id;
      if (sscanf(line, "%lld", &id) == 1) {

        bool is_fresh = false;
        for (int i = 0; i < range_count; i++) {
          if (id >= ranges[i].start && id <= ranges[i].end) {
            is_fresh = true;
            break;
          }
        }
        if (is_fresh) {
          fresh_count++;
        }
      }
    }
  }

  fclose(file);

  printf("Number of fresh ingredient IDs: %d\n", fresh_count);

  return 0;
}

File: day5/part1/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

type Range struct {
	start int64
	end   int64
}

func main() {
	file, err := os.Open("../list.txt")
	if err != nil {
		log.Fatal("Failed to open list.txt: ", err)
	}
	defer file.Close()

	var ranges []Range
	scanner := bufio.NewScanner(file)
	parsingRanges := true
	freshCount := 0

	for scanner.Scan() {
		line := scanner.Text()

		if line == "" {
			parsingRanges = false
			continue
		}

		if parsingRanges {
			parts := strings.Split(line, "-")
			if len(parts) == 2 {
				start, errStart := strconv.ParseInt(parts[0], 10, 64)
				end, errEnd := strconv.ParseInt(parts[1], 10, 64)
				if errStart == nil && errEnd == nil {
					ranges = append(ranges, Range{start: start, end: end})
				}
			}
		} else {
			id, err := strconv.ParseInt(line, 10, 64)
			if err == nil {
				isFresh := false
				for _, r := range ranges {
					if id >= r.start && id <= r.end {
						isFresh = true
						break
					}
				}
				if isFresh {
					freshCount++
				}
			}
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal("Error reading file: ", err)
	}

	fmt.Printf("Number of fresh ingredient IDs: %d\n", freshCount)
}
